training_data_loader.cpp

```diff
@@ -1049,19 +1050,17 @@ std::function<bool(const TrainingDataEntry&)> make_skip_predicate(DataloaderSkip
             // compression ability.
             static constexpr int VALUE_NONE = 32002;
 
-            static constexpr double desired_piece_count_weights[33] = {
-              1.000000, 1.121094, 1.234375, 1.339844, 1.437500, 1.527344, 1.609375,
-              1.683594, 1.750000, 1.808594, 1.859375, 1.902344, 1.937500, 1.964844,
-              1.984375, 1.996094, 2.000000, 1.996094, 1.984375, 1.964844, 1.937500,
-              1.902344, 1.859375, 1.808594, 1.750000, 1.683594, 1.609375, 1.527344,
-              1.437500, 1.339844, 1.234375, 1.121094, 1.000000};
-
-            static constexpr double desired_piece_count_weights_total = []() {
-                double tot = 0;
-                for (auto w : desired_piece_count_weights)
-                    tot += w;
-                return tot;
-            }();
+            // lagrange interpolation weights for desired piece count distribution
+            auto desired_piece_count_weights = [&config](int pc) -> double {
+                double x  = pc;
+                double x1 = 0, y1 = config.pc_y1;
+                double x2 = 16, y2 = config.pc_y2;
+                double x3 = 32, y3 = config.pc_y3;
+                double l1 = (x - x2) * (x - x3) / ((x1 - x2) * (x1 - x3));
+                double l2 = (x - x1) * (x - x3) / ((x2 - x1) * (x2 - x3));
+                double l3 = (x - x1) * (x - x2) / ((x3 - x1) * (x3 - x2));
+                return l1 * y1 + l2 * y2 + l3 * y3;
+            };
```

Comment:
if the filtering performance becomes an issue these could be computed in the capture

Reply:
performance so far is OK, but I agree. It would be nice if they are always consistent with the parameters y1..y3 though.